src/library/features/crates/cratetreemodel.cpp

```diff
@@ -0,0 +1,77 @@
+#include <QVector>
+
+#include "library/trackcollection.h"
+#include "library/features/crates/cratetreemodel.h"
+
+#include "util/db/dbid.h"
+
+namespace {
+
+const CrateId rootId(-1);
+
+} // anonymous namespace
+
+CrateTreeModel::CrateTreeModel(LibraryFeature* pFeature,
+                               TrackCollection* pTrackCollection,
+                               CrateHierarchy* pCrateHierarchy)
+    :m_pFeature(pFeature),
+     m_pTrackCollection(pTrackCollection),
+     m_pCrateHierarchy(pCrateHierarchy) {
+}
+
+bool CrateTreeModel::findParentChildFromPath(Crate& parent,
+                                             Crate& child,
+                                             const QString& idPath) const {
+    QStringList ids = idPath.split("/", QString::SkipEmptyParts);
+
+    // get the last item (child)
+    m_pTrackCollection->crates().readCrateById(CrateId(ids.back()), &child);
+    if (ids.size() > 1) {
+        // get the second to last item (parent)
+        m_pTrackCollection->crates().readCrateById(CrateId(ids.at(ids.size() - 2)), &parent);
+    } else {
+        // if there isn't one return false
+        return false;
+    }
+    return true;
+}
+
+void CrateTreeModel::fillTree(const QStringList& idPaths) {
+    Crate parent, child;
+
+    // looping through the paths we are gsonna get the parent crate and the child crate of
+    // each path (2nd to last and last IDs respectivly). Then we insert the child under the
+    // parent. Since it's sorted alphabetically the parent will always exist before the child
+    for (const auto& idPath : idPaths) {
+        // if there is no parent set the parent as the root of the tree
+        if (!findParentChildFromPath(parent, child, idPath)) {
+            parent.setId(rootId);
+        }
+        if (child.getName() == "") {
```

Comment:
Actually, this shouldn't require a string operation. Would child.getId().isValid() work?